### DISCUSSION TITLE
Entering meta key possible in macOS?

### DISCUSSION THREAD

**Discussion by curiousier-george** [kovidgoyal/kitty#6899] [Q&A]
Is it possible to enter/invoke the meta key from the keyboard in macOS? If not, is there any input (key code?) that would produce the meta key - from Karabiner, for example?

**Comment by Luflosi** [comment]
Is https://sw.kovidgoyal.net/kitty/conf/#opt-kitty.macos_option_as_alt what you're looking for?

**Reply by curiousier-george** [reply]
Thanks for your reply.

I have that set, and I don't think so. Emacs uses the meta key for a lot of things, but alt is different key. It would be more like `macos_option_as_meta`. What I'd really like is `macos_command_as_meta`. But getting the same functionality with the help of Karabiner would also work - is there any way to invoke meta in kitty on macOS?

**Reply by Luflosi** [reply]
Which key is the <kbd>meta</kbd> key if not the <kbd>alt</kbd> key? The most common modifier keys on a Mac are <kbd>Shift</kbd> <kbd>⇧</kbd>, <kbd>Control</kbd> (or <kbd>Ctrl</kbd>) <kbd>⌃</kbd>, <kbd>Option</kbd> (or <kbd>Alt</kbd>) <kbd>⌥</kbd> and <kbd>Command</kbd> (or <kbd>Cmd</kbd>) <kbd>⌘</kbd>. There is also <kbd>Caps Lock</kbd> <kbd>⇪</kbd> but that's almost like <kbd>Shift</kbd> and <kbd>Fn</kbd> which changes the meaning of the function keys and backspace.

**Reply by curiousier-george** [reply]
In emacs, the <kbd>alt</kbd> and <kbd>meta</kbd> modifiers are distinct and can be used separately and/or simultaneously. So if you want to use both, then it isn't natural for the <kbd>meta</kbd> key to be the <kbd>alt</kbd> key. Given the conventions of emacs and the Mac, the <kbd>Command</kbd> key is the most natural key to use for the <kbd>meta</kbd> modifier along with <kbd>Option</kbd> for the <kbd>alt</kbd> modifier. (The GUI version of emacs for the Mac provides special settings for this since it has access to the keycodes and can handle it by itself.)

**Reply by kovidgoyal** [reply]
You need to convince emacs to implement support for the kitty keyboard protocol just like vim, neovim, dte, helix, etc. all do then it will have support for the cmd key out of the box. Or in the meantime use kkp.el

**Comment by curiousier-george** [comment]
Thanks, yes, I'm using kkp.el.

Maybe I'm misunderstanding something - is it possible on a Mac to get kitty to emit all the modifiers (<kbd>shift</kbd>, <kbd>alt</kbd>, <kbd>ctrl</kbd>, <kbd>super</kbd>, <kbd>hyper</kbd>, and <kbd>meta</kbd>) to applications via independent modifier keypresses? That's all I'm trying to do - if that's possible, then everything else has a solution.

**Comment by kovidgoyal** [comment]
On Tue, Dec 12, 2023 at 06:10:04PM -0800, curiousier-george wrote:
> Thanks, yes, I'm using kkp.el.
> 
> Maybe I'm misunderstanding something - is it possible on a Mac to get kitty to emit all the modifiers (<kbd>shift</kbd>, <kbd>alt</kbd>, <kbd>ctrl</kbd>, <kbd>super</kbd>, <kbd>hyper</kbd>, and <kbd>meta</kbd>) to applications via independent modifier keypresses? That's all I'm trying to do - if that's possible, then everything else has a solution.

macs have four modifiers, shift, alt, ctrl and cmd (known as super in
kitty). All of them are supported, as easily demonstrated by running

kitten show-key -m kitty 

These have nothing to do with emacs modifier names, you need to figure
out how to map them to emacs modifiers in emacs, assuming emacs can be
convinced to use the kitty keyboard protocol.

**Comment by curiousier-george** [comment]
Okay, thanks, I understand. What I'd like to able to do is use/emit both alt and meta (in addition to super and hyper) as modifiers to application programs on a Mac, and I'm happy to do any necessary outside-kitty setup to make that happen, but it seems that kitty can't do that on a Mac, is that correct?

(My current setup uses a homegrown analog of the kitty keyboard protocol with Karabiner and iTerm, but it would be so much nicer and easier to be able to use kitty and take advantage of a standard.)

**Comment by kovidgoyal** [comment]
On macs kitty supports only the modifiers macs support, which are the
above four. Though if you want to go nuts you can use send_text in
kitty.conf to emit any arbitrary bytes you like on any key presses.

**Reply by curiousier-george** [reply]
Okay, thanks. No, I don't want to go nuts again (been there/done that/still works/kind of defeats to purpose of using kkp).

What I was hoping was that there would be some way to simulate keypresses (using Karabiner, for example) to induce kitty to emit any possible modifier combination to applications using the kitty keyboard protocol.

One conception of it could be that kitty supported virtual physical keys (e.g., meta) such that if Karabiner sent them to kitty (even though they didn't exist physically), kitty would be able to act like they were pressed and emit the corresponding modifier (meta, for example) to programs running in the terminal.

The way I currently do it involves treating left and right modifiers differently and also using the Mac function key as a modifier.

Thanks for your help.
